### PR TITLE
Building on Jun's previous changes

### LINKED ISSUE #21: implement mini_batch_gradient_descent

### LINKED ISSUE #22: update sklearn/linear_model/_sgd_fast.pyx::_plain_sgd to support mini_batch_gradient_descent

### PR DESCRIPTION
**Pull request by RylandWang**
#### Reference Issues/PRs

Addresses issue 21

#### What does this implement/fix? Explain your changes.
Applied weight update on entire mini-batch of vectors instead of just the last single vector.

### CODE REVIEW

File: sklearn/linear_model/_sgd_fast.pyx [hunk 1]
```diff
@@ -635,13 +642,14 @@ def _plain_sgd(np.ndarray[double, ndim=1, mode='c'] weights,
...
                 # TODO: not sure about this part
                 if update != 0.0:
-                    # TODO: add a for loop, loop through the  batch_data_pointers , batch_index_pointers, xnnzs,
-                    # update the w
-                    w.add(x_data_ptr, x_ind_ptr, xnnz, update)
-                    with gil:
-                        print("line 621: update != 0, weights", weights)
-                    if fit_intercept == 1:
-                        intercept += update * intercept_decay
+                    # traverse data points in current batch and apply weight update
```
Comment (linjun9) [inline sklearn/linear_model/_sgd_fast.pyx:638]: yep, this is correct, but need to address a case where that `x_data` might be in validation_set, we need to skip

Comment (VishwaP98) [reply sklearn/linear_model/_sgd_fast.pyx:638]: I am doing that in my PR @linjun9

File: sklearn/linear_model/_sgd_fast.pyx [hunk 2]
```diff
@@ -520,7 +523,7 @@ def _plain_sgd(np.ndarray[double, ndim=1, mode='c'] weights,
         optimal_init = 1.0 / (initial_eta0 * alpha)
 
     t_start = time()
-    print("batch_size", batch_size)
+    print("batch_size1000", batch_size)
```
Comment (linjun9) [inline sklearn/linear_model/_sgd_fast.pyx]: remove this line please, or you can put it to a `if verbose block`

File: sklearn/linear_model/_sgd_fast.pyx [hunk 3]
```diff
@@ -596,10 +599,14 @@ def _plain_sgd(np.ndarray[double, ndim=1, mode='c'] weights,
...
+                    # go to next element
+                    batch_counter += 1
+
                 # outside of the batch loop ---------------
-                avg_batch_loss = sum_batch_loss / (batch_size - validation_sample_count)
-                avg_batch_dloss = sum_batch_dloss / (batch_size - validation_sample_count)
-                # batch_x_sum.scale(1 / (batch_size - validation_sample_count))
+                avg_batch_loss = sum_batch_loss / (batch_size)
+                avg_batch_dloss = sum_batch_dloss / (batch_size)
+                weight_update = weight_update / batch_size
```
Comment (linjun9) [inline sklearn/linear_model/_sgd_fast.pyx:608]: I see what you are trying to do here. But if we want to multiply `update` with `avg_weight_update` for you should reset `weight_update` before entering to the `batch loop`

File: sklearn/linear_model/demo.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import numpy as np
```
Comment (linjun9) [inline sklearn/linear_model/demo.py]: please remove this demo file, we can put it in a gist file and add that to our report

File: sklearn/linear_model/_sgd_fast.pyx (hunk 2 above)
Comment (VishwaP98) [reply sklearn/linear_model/_sgd_fast.pyx]: done

File: sklearn/linear_model/demo.py (hunk 4 above)
Comment (VishwaP98) [reply sklearn/linear_model/demo.py]: done

File: sklearn/linear_model/_sgd_fast.pyx (hunk 3 above)
Comment (VishwaP98) [reply sklearn/linear_model/_sgd_fast.pyx:608]: done

File: sklearn/linear_model/_sgd_fast.pyx [hunk 5]
```diff
@@ -596,10 +602,14 @@ def _plain_sgd(np.ndarray[double, ndim=1, mode='c'] weights,
...
```
Comment (linjun9) [inline sklearn/linear_model/_sgd_fast.pyx:608]: I was gonna say if we do want to calculate the weight for each of the data pointers independently, we can do it in the `for loop` when we calculate our `w`. But I'm ok with taking the avg_weight

**Review by linjun9 — COMMENTED** [review]
I think I'm ok with what we have so far, can you do a final walk-through, delete those print comments, and commented code.

**Pull request merged**